Add STM32H7 Support

iabdalkader opened the pull request.

iabdalkader committed:
Add H743 board files.

iabdalkader committed:
Add H7 Makefile flags.

iabdalkader committed:
Add flash support for H7

iabdalkader committed:
Add H7 system initialization.

iabdalkader committed:
Disable NUCLEO_H743ZI ADC.

iabdalkader committed:
Fix H7 cache enable.

iabdalkader committed:
Fix H7 board config typo.

iabdalkader committed:
Fix H7 pin_high/low macros.

iabdalkader committed:
Remove error_handler from system_stm32.c

iabdalkader committed:
Fix NUCLEO_H743ZI PLLP

iabdalkader committed:
Add H7 WDT support.

iabdalkader committed:
Add H7 UART support.

iabdalkader committed:
Add H7 timer support.

iabdalkader committed:
Add modmachine H7 support.

iabdalkader committed:
Fix H7 OTG AF.

iabdalkader committed:
Add H7 SWS_PLL1 flag

iabdalkader committed:
Add H7 DMA support.

iabdalkader committed:
Fix H7 flash support.

iabdalkader committed:
Update Makefile for H7.

* No strict aliasing (or fix HAL errors)
* Add fdcan.c instead of can.c

iabdalkader committed:
Add H7 RTC support.

iabdalkader committed:
Add H7 extint support.

iabdalkader committed:
Fix fdcan/can prefix.

iabdalkader committed:
Update H7 startup code.

iabdalkader committed:
Update H7 linker script.

iabdalkader committed:
Fix H7 USB PLL config.

iabdalkader committed:
AF is already set.

iabdalkader committed:
Fix H7 linker script FS sector size.

iabdalkader committed:
Update H7 board config.

* Enable USB.
* Add LEDs.

iabdalkader committed:
Add H7 board init.

iabdalkader committed:
Fix H7 flash text address.

iabdalkader committed:
Fix H7 LED pins.

iabdalkader committed:
Adjust H7 heap end.

iabdalkader committed:
Enable H7 UART_REPL.

iabdalkader committed:
Add H7 UART pins.

iabdalkader committed:
Add STM32H7 common configuration.

iabdalkader committed:
Remove H7 unused APBPrescTable.

iabdalkader committed:
Enable the H7 I/O Compensation Cell.

rolandvs commented:
@iabdalkader This is fantastic! I'm going to try this on my NUCLEO-H743 board and another board I have here.

rolandvs commented:
Ran into some problems when:  `make BOARD=NUCLEO_H743ZI`. Most of it compiles.

```
./../lib/stm32lib/CMSIS/STM32H7xx/Include/stm32h743xx.h:2355:45: error: expected ')' before '*' token
 #define SDMMC               ((SDMMC_TypeDef *) SDMMC_BASE)
                                             ^
./pin_defs_stm32.h:128:20: note: in expansion of macro 'SDMMC'
     SDMMC_TypeDef *SDMMC
                    ^~~~~
./pin.h:46:5: note: in expansion of macro 'PIN_DEFS_PORT_AF_UNION'
     PIN_DEFS_PORT_AF_UNION
     ^~~~~~~~~~~~~~~~~~~~~~
../../lib/stm32lib/CMSIS/STM32H7xx/Include/stm32h743xx.h:2355:48: error: expected ')' before 'SDMMC_BASE'
 #define SDMMC               ((SDMMC_TypeDef *) SDMMC_BASE)
```

I took out the `void  *SDMMC; \` and compiling continued w/o error, however linking gave a new error:

```
LINK build-NUCLEO_H743ZI/firmware.elf
build-NUCLEO_H743ZI/stm32_it.o: In function `SysTick_Handler':
stm32_it.c:(.text.SysTick_Handler+0x30): undefined reference to `uwTick'
build-NUCLEO_H743ZI/systick.o: In function `mp_hal_delay_ms':
systick.c:(.text.mp_hal_delay_ms+0x40): undefined reference to `uwTick'
build-NUCLEO_H743ZI/systick.o: In function `mp_hal_ticks_ms':
systick.c:(.text.mp_hal_ticks_ms+0x8): undefined reference to `uwTick'
make: *** [build-NUCLEO_H743ZI/firmware.elf] Error 1
```



Running gcc on OSX...

iabdalkader commented:
https://github.com/micropython/micropython/issues/3515#issuecomment-368249383

dpgeorge commented on ports/stm32/boards/NUCLEO_H743ZI/mpconfigboard.h:
```diff
@@ -0,0 +1,84 @@
+*/
+#define MICROPY_HW_CLK_PLLM (4)
+#define MICROPY_HW_CLK_PLLN (400)
+#define MICROPY_HW_CLK_PLLP (4)
```
This should probably be (2) to get 400MHz CPU freq.

dpgeorge commented on ports/stm32/boards/NUCLEO_H743ZI/stm32h7xx_hal_conf.h:
```diff
@@ -0,0 +1,434 @@
+  *        (when HSE is used as system clock source, directly or through the PLL).
+  */
+#if !defined  (HSE_VALUE)
+#define HSE_VALUE    ((uint32_t)25000000) /*!< Value of the External oscillator in Hz */
```
The NUCLEO F743ZI board that I have has an 8MHz external crystal (HSE) so I think this needs to be changed to 8000000.

dpgeorge replied:
Oops, I meant NUCLEO H743ZI board.

dpgeorge commented:
Re the USB: I'm confused as to which USB peripheral is connected to which pins.  The AF's are defined like this:
```
Inc/stm32h7xx_hal_gpio_ex.h:#define GPIO_AF10_OTG2_HS       ((uint8_t)0xA)  /* OTG2_HS Alternate Function mapping */
Inc/stm32h7xx_hal_gpio_ex.h:#define GPIO_AF10_OTG1_FS       ((uint8_t)0xA)  /* OTG1_FS Alternate Function mapping */
```
yet the peripheral pointers are defined like this:
```
Include/stm32h743xx.h:#define USB_OTG_HS                   USB1_OTG_HS
Include/stm32h743xx.h:#define USB_OTG_FS                   USB2_OTG_FS
```
It could just be a typo in the AF10 definitions, so then USB2 would be the FS one.  Apart from this, the USB clock looks correct, so I'm not sure why it doesn't work.

kwagyeman commented:
Great progress folks!

iabdalkader committed:
Fix H7 HAL HSE_VALUE

iabdalkader committed:
Fix H7 PLLP value.

iabdalkader committed:
Set UART clock sources.

iabdalkader commented:
I fixed HSE_VALUE (although I don't really think it's used) and I fixed PLLP. Now the UART REPL baudrate is as expected (115200). USB still not working.

iabdalkader commented:
@dpgeorge I was finally able to narrow down the issue. Something breaks in REPL..If you comment the following lines in main.c:

```
+        //if (pyexec_mode_kind == PYEXEC_MODE_RAW_REPL) {
+        //    if (pyexec_raw_repl() != 0) {
+        //        break;
+        //    }
+        //} else {
+        //    if (pyexec_friendly_repl() != 0) {
+        //        break;
+        //    }
+        //}

```

And **disable REPL UART** USB will work (CDC + MSC)

dpgeorge commented:
@iabdalkader following on from your work above, I narrowed it further to a problem with WFI: if WFI is executed then it puts the USB peripheral to sleep and that breaks it.  To fix you need to allow the USB to operate during a WFI (the following worked for me but maybe it could be simplified since we don't need both the USB1 and USB2 peripherals, and the ULPI is not used(?)):
```diff
--- a/ports/stm32/usbd_conf.c
+++ b/ports/stm32/usbd_conf.c
@@ -93,6 +93,11 @@ void HAL_PCD_MspInit(PCD_HandleTypeDef *hpcd)
     HAL_GPIO_Init(GPIOA, &GPIO_InitStruct); 
 #endif
 
+    __HAL_RCC_USB1_OTG_HS_CLK_SLEEP_ENABLE();
+    __HAL_RCC_USB1_OTG_HS_ULPI_CLK_SLEEP_DISABLE();
+    __HAL_RCC_USB2_OTG_FS_CLK_SLEEP_ENABLE();
+    __HAL_RCC_USB2_OTG_FS_ULPI_CLK_SLEEP_DISABLE();
+
     /* Enable USB FS Clocks */ 
     __USB_OTG_FS_CLK_ENABLE();
```
But even with this patch it still seems that the USB is broken when UART REPL is enabled.  So you still need to disable UART REPL.  Probably it has something to do with interrupts, maybe with UART REPL enabled there's an IRQ which is not being serviced and is calling over and over again, starving the USB interrupt.

iabdalkader commented:
@dpgeorge Awesome! Thank you for help. I spent hours on this without a debugger (I even patched stlink to support H7 but breakpoints are not working). I don't think ULPI or HS is needed but maybe we should keep both just in case the code is used on a different board (will use  `#MICROPY_HW_USB_XX`). Will look into the UART REPL issue, but I could disable it in mpconfigboard.h until it's fixed.

dpgeorge commented:
> I spent hours on this without a debugger (I even patched stlink to support H7 but breakpoints are not working)

The latest OpenOCD (built from source) has support for the H7.  I was actually just using direct USB DFU mode to flash the H7 MCU.

iabdalkader committed:
* Disable USB clock sleep.

iabdalkader commented:
@dpgeorge Should the USB sleep be disabled for all MCUs or just the H7 ? I just noticed it's not specific to H7, and it's already enabled in HS mode for all MCUs.

dpgeorge commented:
> Should the USB sleep be disabled for all MCUs or just the H7 ?

It's probably enabled by default on all existing F4/L4's (by enabled I mean that the USB clock will still run during WFI) and that's why they work without doing anything.  On the F733 I needed to enable it explicitly for the HS peripheral.

I think just add a H7 specific piece of code, which can be generalised later to other MCUs.

iabdalkader committed:
Disable OTG FS clock sleep for H7.

iabdalkader commented:
@dpgeorge 
> The latest OpenOCD (built from source) has support for the H7. 

It's not working either.

iabdalkader committed:
Enable H7 D2 SRAM1/2/3 clocks.

iabdalkader committed:
H7 AXI SRAM workaround (see Errata 2.2.15).

iabdalkader committed:
Set H7 UARTs clock source in init0.

iabdalkader committed:
Fix uart clock source init.

iabdalkader commented:
@dpgeorge I fixed the uart issue, usb and uart are both working now.

iabdalkader committed:
Remove AXI workaround from main.c

* Already set in system_stm32.c

iabdalkader committed:
Disable Wait for PWR_FLAG_VOSRDY.

* F7 gets stuck here.

iabdalkader commented:
@dpgeorge When do you think you can review and merge this ? Or should I be doing something else first ?

BTW, while working on system_stm32.c I noticed this, it doesn't look right but maybe I'm missing something ?

https://github.com/micropython/micropython/blob/master/ports/stm32/system_stm32.c#L113
https://github.com/micropython/micropython/blob/master/ports/stm32/system_stm32.c#L127

dpgeorge commented:
> BTW, while working on system_stm32.c I noticed this, it doesn't look right but maybe I'm missing something ?

Yes that does look like a bug.  Will need to look at the history of that line to see how it came to be.  But I'll leave it alone for now.

> When do you think you can review and merge this ? Or should I be doing something else first ?

I'll need to review it fully.

One thing that might be good to do is have separate system_stm32.c and startup_stm32.S files (and maybe even stm32_it.c) for each MCU series.  Especially the startup_stm32.S file, it's already pretty messy trying to support F4, F7 and L4.

iabdalkader commented:
> One thing that might be good to do is have separate system_stm32.c and startup_stm32.S files (and maybe even stm32_it.c) for each MCU series. Especially the startup_stm32.S file, it's already pretty messy trying to support F4, F7 and L4.

@dpgeorge I couldn't agree more, startup_stm32.S is very confusing. I'm already using separate startup files in OpenMV's repo (and we only support 3 boards). Should this be done now or after merging ?

dpgeorge commented:
> Should this be done now or after merging ?

Doing it now would be better.  Maybe put it in boards/, call it boards/startup_stm32h7.S.

iabdalkader committed:
Revert startup_stm32.S.

iabdalkader committed:
Add H7 series startup file.

iabdalkader committed:
Add F7 series startup file.

iabdalkader committed:
Add F4 series startup file.

iabdalkader committed:
Add L4 series startup file.

iabdalkader committed:
Use separate startup file for each MCU series.

iabdalkader commented:
@dpgeorge It's done. I don't think system should be split it's not that bad. One thing I wanted to do for a very long time now, is to move all the init0 functions (and maybe init functions?) to  their own section in linker script, and call them in a loop (kernel style __init).

dpgeorge commented:
> I don't think system should be split it's not that bad.

It seems quite different for F0 and F3.  But let's leave that for a later date.

> One thing I wanted to do for a very long time now, is to move all the init0 functions (and maybe init functions?) to their own section in linker script, and call them in a loop (kernel style __init).

Can you please elaborate on this?  Some of the init0 (and all the init) need to be called on a soft reset.

iabdalkader commented:
> Can you please elaborate on this? Some of the init0 (and all the init) need to be called on a soft reset.

It's basically just placing function pointers to init/init0 functions in a linker script section, and calling them in a loop in main (using say `_init_start=.` and `_init_end=.` labels and a function pointer). The same can be done for `deinit` functions and more than one section (of function pointers) can be used for say early init, subsytem init etc.. when order matters. The overhead is 4 bytes per init function. The reason for this is main() is getting messy with all the init/init0 functions (>20 functions) and will add more for H7, and generally it's just a nice way to call init functions.
See https://linuxgazette.net/157/amurray.html

kwagyeman commented:
Hi all, any progress on this?

dpgeorge commented:
Thanks @iabdalkader for all your hard work on this.  I've now gone through and checked everything (it seems good) and done a fair bit of rebasing and squashing of commits to clean it up.  I updated the stm32lib to include the required mods for H7.  These changes have now been merged, between commits a3c721772eea87744d49ee65b9dd157f28bbcb76 and eb56efb4347265678c90bd67a8ea3c89b1d052eb

Excellent!

dpgeorge commented:
> It's basically just placing function pointers to init/init0 functions in a linker script section, and calling them in a loop in main

Ok, thanks for the explanation.  If you want to make a PR for it feel free, but it won't be a priority for me right now.

iabdalkader commented:
> Ok, thanks for the explanation. If you want to make a PR for it feel free, but it won't be a priority for me right now.

Will send a PR for review later.

rolandvs commented:
@dpgeorge @iabdalkader I was away on holidays for a week and the world has changed :-). Great just uploaded the code to my NUCLEO-H743ZI board and have REPL running!

rolandvs commented:
@iabdalkader The port is working well. However I have some questions/remarks. As they are related I put them together although a separate issue per item would be preferred way to work?!

Changing the crystal like `25MHz` for the `HSE_VALUE` will not only require changes to the variables of the PLL for the CPU speed, but also some changes to the other PLL for the USB port that is buried in `system_stm32.c`. Should be bring out those variables to the `mpconfigboard.h` as well?

rolandvs commented:
The flash size is 2048K for most derivates of the H7xx, however in the `flash.c` I have added some more segments to be able to reach all of them:

```
#elif defined(STM32H7)

/** flash memory is 1024K-2048K **/
#if 0
static const flash_layout_t flash_layout[] = {
    { 0x08000000, 0x20000, 8 },
};
#else
static const flash_layout_t flash_layout[] = {
    { 0x08000000, 0x20000, 16 },
};
#endif
```

rolandvs commented:
I have increased the flash drive size. I did so by adding the following in `flashbdev.c` :

```
#elif defined(STM32H743xx)

// The STM32H743 flash sectors are 128K
#define CACHE_MEM_START_ADDR (0x20000000) // DTCM data RAM, 128k
#define FLASH_SECTOR_SIZE_MAX (0x20000) // 128k max
#define FLASH_MEM_SEG1_START_ADDR (0x08020000) // sector 1
#define FLASH_MEM_SEG1_NUM_BLOCKS (256) // Sector 1: 128k / 512b = 256 blocks

// added for more flash storage
#if MICROPY_HW_FLASH_EXTRA
#define FLASH_MEM_SEG2_START_ADDR (0x08040000) // starts right after first block segment
#define FLASH_MEM_SEG2_NUM_BLOCKS (1536)  // 6*128K/512 blocks
#endif
```

Also made the `boards/stm32h743.ld` local to the board implementation, which seems more logical as the memory use is depending on the board (and not as a generic MCU).
In this case the memory use in the linker file is like:
```
MEMORY
{
	/* flash is dual bank 1024K, however used as a single 2048K */
    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 2048K	  /* sectors 0..15 of 128K = 2028K */
    FLASH_ISR (rx)  : ORIGIN = 0x08000000, LENGTH = 128K    /* sector  0     of 128K =  128K */
    FLASH_FS (r)    : ORIGIN = 0x08020000, LENGTH = 128K    /* sector  1...7 of 128K =  896K */
    FLASH_TEXT (rx) : ORIGIN = 0x08100000, LENGTH = 1024K   /* sectors 8..15 of 128K = 1024K */

    DTCM (xrw)      : ORIGIN = 0x20000000, LENGTH = 128K    /* used for storage cache */
    RAM (xrw)       : ORIGIN = 0x24000000, LENGTH = 512K    /* AXI SRAM */
}
```

Of course the `mpconfigboard.mk` has been adapted and looks like:
```
MCU_SERIES = h7
CMSIS_MCU = STM32H743xx
MICROPY_FLOAT_IMPL = double
AF_FILE = boards/stm32h743_af.csv
LD_FILES = boards/NUCLEO_H743ZI/stm32h743.ld boards/common_ifs.ld
TEXT0_ADDR = 0x08000000
TEXT1_ADDR = 0x08100000
```
